core/state-machine/src/changes_trie/surface_iterator.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The best way to understand how this iterator works is to imagine some 2D terrain that have some mountains
+//! (digest changes tries) and valleys (changes tries for regular blocks). There are gems (blocks) beneath the
+//! terrain. Given the request to find all gems in the range [X1; X2] this iterator will return **minimal set**
+//! of points at the terrain (mountains and valleys() inside this range that have to be drilled down to
+//! search for gems.
+
```

Comment:
Nope - drilldown will 'drill' where surface where surface iterator points. BTW - if you think you could provide better documentation for this iterator, I will gladly accept it :) I can't think of better way than to provide this analogy :)

Reply:
Drilldown is using surface, and I thought the analogy refers more to the combination of both iterators (and since it is build from drilldown code it could make sense). I will try to think at how to document this, but I am usually not really good at that, I find the analogy interesting :)